text/0000-ptr-meta.md

```diff
@@ -0,0 +1,398 @@
+  (similar to how `Sized` and `Unsize` are implemented automatically).
+* A `Thin` [trait alias].
+  If this RFC is implemented before type aliases are,
+  uses of `Thin` should be replaced with its definition.
+* A `metadata` free function
+* A `from_raw_parts` constructor for each of `*const T` and `*mut T`
+
+The bounds on `null()` and `null_mut()` function in that same module
+as well as the `NonNull::dangling` constructor
+are changed from (implicit) `T: Sized` to `T: ?Sized + Thin`.
+Similarly for the `U` type parameter of the `NonNull::cast` method.
+This enables using those functions with [extern types].
+
+The `Pointee` trait is implemented for all types.
+This can be relied on in generic code,
+even if a type parameter `T` does not have an explicit `T: Pointee` bound.
+This is similar to how the `Any` trait can be used without an explicit `T: Any` bound,
+only `T: 'static`, because a blanket `impl<T: 'static> Any for T {…}` exists.
+(Except that `Pointee` is not restricted to `'static`.)
+
+For the purpose of pointer casts being allowed by the `as` operator,
+a pointer to `T` is considered to be thin if `T: Thin` instead of `T: Sized`.
+This similarly includes extern types.
+
+`std::raw::TraitObject` and `std::raw` are deprecated and eventually removed.
+
+[trait alias]: https://github.com/rust-lang/rust/issues/41517
+[extern types]: https://github.com/rust-lang/rust/issues/43467
+
+```rust
+/// This trait is automatically implement for every type.
+///
+/// Raw pointer types and reference types in Rust can be thought of as made of two parts:
+/// a data pointer that contains the memory address of the value, and some metadata.
+///
+/// For statically-sized types that implement the `Sized` traits,
+/// pointers are said to be “thin”, metadata is zero-sized, and metadata’s type is `()`.
+///
+/// Pointers to [dynamically-sized types][dst] are said to be “fat”
+/// and have non-zero-sized metadata:
+///
+/// * For structs whose last field is a DST, metadata is the metadata for the last field
+/// * For the `str` type, metadata is the length in bytes as `usize`
+/// * For slice types like `[T]`, metadata is the length in items as `usize`
+/// * For trait objects like `dyn SomeTrait`, metadata is [`&'static VTable`][VTable]
+///
+/// In the future, the Rust language may gain new kinds of types
+/// that have different pointer metadata.
+///
+/// Pointer metadata can be extracted from a pointer or reference with the [`metadata`] function.
+/// The data pointer can be extracted by casting a (fat) pointer
+/// to a (thin) pointer to a `Sized` type the `as` operator,
+/// for example `(x: &dyn SomeTrait) as *const SomeTrait as *const ()`.
+///
+/// [dst]: https://doc.rust-lang.org/nomicon/exotic-sizes.html#dynamically-sized-types-dsts
+#[lang = "pointee"]
+pub trait Pointee {
+    /// The type for metadata in pointers and references to `Self`.
+    type Metadata: Copy + Send + Sync + Ord + Hash + 'static;
```

Comment:
Sorry I missed the Pin API. Could you also add `Unpin` to the list?

Reply:
It’s somewhat worrying if this list of bounds is growing. Could there be another trait in the future that we’d want to add here, but can’t because the trait is stable?

(In an hypothetical custom-DST future where user-defined impls can exist. In this RFC I think they cannot because they would always overlap with impls provided by the compiler through "magic".)

Reply:
What are the consequences for generic code not to be able to assume `T::Metadata: Unpin` specifically?

Reply:
Can types even be `Copy` but not `Unpin`? Can we add `Unpin` as a super trait for `Copy`, in a backwards compatible way? I think so, because `Unpin` is an auto trait.

Reply:
@plietar https://doc.rust-lang.org/std/marker/struct.Pinned.html

Reply:
Added.